### PR TITLE
New phase field cracking material that modifies existing stress

### LINKED ISSUE #11685: Create phase field cracking material that can be added on to an existing stress
## Rationale

Currently, each time we want to add the phase field crack model to a new constitutive model we have to create a new material that computes the stress and then modifies it to add cracking. This is cumbersome and promotes code duplication.

## Description

It would be much easier if we could use an existing material to calculate the stress and then just modify that stress to add cracking

## Impact

This is new capability and won't change any existing tests or capabilities

### PR DESCRIPTION
**Pull request by tonkmr**
I added a new material that implements the phase field cracking model by modifying the existing stress rather than computing the stress and modifying it in the same material. I also improved the Jacobian in the existing implementation of the anisotropic fracture model. 

@jiangwen84 please review.

Closes #11685.

### CODE REVIEW

**Comment by tonkmr** [comment]
@dschwen I would like to modify the TensorMechanicsAction to automatically know which base_name to use for output for the stresses and strains, but I am having trouble figuring out how to accomplish that. Any thoughts?

**Comment by tonkmr** [comment]
I don't think the app test failure was caused by these changes.

**Comment by permcody** [comment]
Agreed - @YaqiWang - Have you seen that test fail before?

**Comment by jiangwen84** [comment]
@tonkmr Do you think we still need to keep ```ComputeLinearElasticPFFractureStress```? It seems the ```ComputeCrackedStress``` can be used for linear elastic materials too.

**Comment by YaqiWang** [comment]
Not for a while. But since it is from cpr, I am not very surprised. I guess invalidating it will possible fix.

**Comment by tonkmr** [comment]
@jiangwen84, I think we do not need ComputeLinearElasticPFFractureStress long term, but I didn't want to eliminate it right away. People may be still using it and there may be more issues we need to address with ComputeCrackedStress. Still, I would be open to deleting it if we want to.

File: modules/tensor_mechanics/include/materials/ComputeCrackedStress.h [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  /// Use current value of history variable
+  bool _use_current_hist;
+
+  const MaterialProperty<RankTwoTensor> & _mechanical_strain;
+
+  /// Uncracked stress calculated by another material and modified here to include cracks
+  const MaterialProperty<RankTwoTensor> & _uncracked_stress;
+
+  /// Uncracked Jacobian_mult calculated by another material and modified Here
+  const MaterialProperty<RankFourTensor> & _uncracked_Jacobian_mult;
```
Comment (jiangwen84) [inline modules/tensor_mechanics/include/materials/ComputeCrackedStress.h]: Did you modify the _uncracked_Jacobian_mult? If not, we should delete 'modified here' from the comment.

File: modules/tensor_mechanics/src/materials/ComputeCrackedStress.C [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  if (G0_pos > _hist_old[_qp])
+    _hist[_qp] = G0_pos;
+  else
+    _hist[_qp] = _hist_old[_qp];
+
+  Real hist_variable = _hist_old[_qp];
+  if (_use_current_hist)
+    hist_variable = _hist[_qp];
+
+  // Compute interpolation function and derivatives
```
Comment (jiangwen84) [inline modules/tensor_mechanics/src/materials/ComputeCrackedStress.C]: I would call it degradation function in the pf fracture context.

File: modules/tensor_mechanics/src/materials/ComputeLinearElasticPFFractureStress.C [hunk 3]
```diff
@@ -46,62 +61,59 @@ ComputeLinearElasticPFFractureStress::computeQpStress()
...
+  if (G0_pos > _hist_old[_qp])
+    _hist[_qp] = G0_pos;
+  else
+    _hist[_qp] = _hist_old[_qp];
+
+  Real hist_variable = _hist_old[_qp];
+  if (_use_current_hist)
+    hist_variable = _hist[_qp];
+
+  // Compute interpolation function and derivatives
```
Comment (jiangwen84) [inline modules/tensor_mechanics/src/materials/ComputeLinearElasticPFFractureStress.C]: same here, degradation function

**Review by jiangwen84 — COMMENTED** [review]
It looks good to me. I made a few minor comments. Beside, I wonder if it is worth creating a base class from which the ```ComputeCrackedStress```, ```ComputeLinearElasticPFFractureStres``` and ```ComputeIsotropicLinearElasticPFFractureStress``` derive to avoid code and variable duplications.

**Review by jwpeterson — CHANGES_REQUESTED** [review]
Just updating the review status of the PR.

Please address @jiangwen84's comments.

File: modules/tensor_mechanics/src/materials/ComputeCrackedStress.C [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  RankTwoTensor eigvec;
+  RankFourTensor Ppos = _uncracked_stress[_qp].positiveProjectionEigenDecomposition(eigval, eigvec);
+  RankFourTensor Pneg = I4sym - Ppos;
+
+  // Project the positive and negative stresses
+  RankTwoTensor stress0pos = Ppos * _uncracked_stress[_qp];
+  RankTwoTensor stress0neg = Pneg * _uncracked_stress[_qp];
+
+  // Compute the positive and negative elastic energies
+  Real G0_pos = (stress0pos).doubleContraction(_mechanical_strain[_qp]) / 2.0;
```
[the remark this answers is not in the document]
Comment (tonkmr) [reply modules/tensor_mechanics/src/materials/ComputeCrackedStress.C]: Now that I think about this, shouldn't it be the elastic strain? The elastic strain dotted into the elastic stress is the elastic energy, which is what is causing the brittle fracture. @jiangwen84 what do you think? 

In Pritam's HyperElasticPhaseFieldIsoDamage I think he uses the elastic energy (he calls it _ee).

**Comment by tonkmr** [comment]
@jiangwen84 I am currently using FiniteStrainPlasticMaterial in my plastic test. Is there a better plastic material that I should be using instead? The solve doesn't seem very robust in that material. I just want what is considered the best yield stress based model.

**Comment by jiangwen84** [comment]
@tonkmr , did you try ```IsotropicPlasticityStressUpdate``` together with ```ComputeMultipleInelasticStress```. One test can be found:  tensor_mechanics/test/tests/recompute_radial_return/isotropic_plasticity_finite_strain.i

**Comment by tonkmr** [comment]
@jiangwen84 those files worked better, thanks. I changed the plasticity test to use those materials. I ran it to full extension of the crack and it worked.

File: modules/tensor_mechanics/include/materials/ComputeCrackedStress.h [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+  /// Stress being computed by this kernel
+  MaterialProperty<RankTwoTensor> & _stress;
+
+  MaterialProperty<Real> & _F;
+  MaterialProperty<Real> & _dFdc;
+  MaterialProperty<Real> & _d2Fdc2;
+  MaterialProperty<RankTwoTensor> & _d2Fdcdstrain;
+  MaterialProperty<RankTwoTensor> & _dstress_dc;
+
```
[the remark this answers is not in the document]
Comment (tonkmr) [reply modules/tensor_mechanics/include/materials/ComputeCrackedStress.h]: I added the comment

**Comment by tonkmr** [comment]
@jwpeterson I think that this may be ready to be merged. @jiangwen84 seems happy with the current state.

**Review by jwpeterson — APPROVED** [review]

**Pull request merged**